lib/ip_list.go

```diff
@@ -1,250 +1,282 @@
-// line.
-func IPListFromStdin() []net.IP {
-	return IPListFromReader(os.Stdin)
-}
-
-// IPListFromFile returns a list of IPs found in a file.
-func IPListFromFile(pathToFile string) ([]net.IP, error) {
-	f, err := os.Open(pathToFile)
-	if err != nil {
-		return nil, err
-	}
-
-	return IPListFromReader(f), nil
-}
-
-// IPListFromFiles returns a list of IPs found in a list of files.
-func IPListFromFiles(paths []string) (ips []net.IP, err error) {
-	// collect IPs lists together first, then allocate a final list and do
-	// a fast transfer.
-	ipLists := make([][]net.IP, len(paths))
-	totalIPs := 0
-	for i, p := range paths {
-		ipLists[i], err = IPListFromFile(p)
-		if err != nil {
-			return nil, err
-		}
-		totalIPs += len(ipLists[i])
-	}
-
-	ips = make([]net.IP, 0, totalIPs)
-	for _, ipList := range ipLists {
-		ips = append(ips, ipList...)
-	}
-
-	return ips, nil
-}
+package lib
+
+import (
+	"bufio"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"strings"
+)
+
+// isFQDN returns true if domain is a fully qualified domain
+// name. The logic comes from main.go
+func isFQDN(domain string) bool {
+	return len(domain) >= 3 && strings.IndexByte(domain, '.') != -1
+}
+
+// IPListFrom returns a list of IPs from stdin and a list of inputs which is
+// interpreted to contain IPs, IP ranges, IP CIDRs and files with IPs in them,
+// all depending upon which flags are set.
+func IPListFrom(
+	inputs []string,
+	stdin bool,
+	ip bool,
+	iprange bool,
+	cidr bool,
+	file bool,
+) ([]net.IP, error) {
+	ips := make([]net.IP, 0, len(inputs))
+
+	// prevent edge cases with all flags turned off.
+	if !stdin && !ip && !iprange && !cidr && !file {
+		return ips, nil
+	}
+
+	// start with stdin.
+	if stdin {
+		stat, _ := os.Stdin.Stat()
+
+		isPiped := (stat.Mode() & os.ModeNamedPipe) != 0
+		isTyping := (stat.Mode()&os.ModeCharDevice) != 0 && len(inputs) == 0
+
+		if isTyping {
+			fmt.Println("** manual input mode **")
+			fmt.Println("Enter all IPs, one per line:")
+		}
+
+		if isPiped || isTyping || stat.Size() > 0 {
+			ips = append(ips, IPListFromStdin()...)
+		}
+	}
+
+	// parse `inputs`.
+	for _, input := range inputs {
+		if iprange {
+			_ips, err := IPListFromRangeStr(input)
+			if err == nil {
+				ips = append(ips, _ips...)
+				continue
+			}
+		}
+
+		if ip && StrIsIPStr(input) {
+			ips = append(ips, net.ParseIP(input))
+			continue
+		}
+
+		if cidr && StrIsCIDRStr(input) {
+			_ips, _ := IPListFromCIDR(input)
+			ips = append(ips, _ips...)
+			continue
+		}
+
+		if file && FileExists(input) {
+			_ips, err := IPListFromFile(input)
+			if err != nil {
+				return nil, err
+			}
+			ips = append(ips, _ips...)
+			continue
+		}
+
+		// Use FQDN here
+		// According to main.go, this is how
+		// you know a string is a FQDN
+		// This goes AFTER the file so the cli does not confuse
+		// files like ips.txt as domain names
+		if ip && isFQDN(input) {
+			// Get the IP address(es) of the domains
+			inputIps, err := net.LookupIP(input)

```

Comment:
So usually the bulk command is meant to be used with a large amount of inputs. The methodology used here means that we'll be looking up the domains via DNS one-by-one, which is going to be extremely slow and bottleneck the entire pipeline.

I'm actually not sure what the right answer is here. We could solve it by going all-out on concurrency with these, but is it really worth the effort?

Reply:
Could we use a Wait Group? Create an array of only domain names and, if larger than a certain arbitrary number, use a wait group to solve them in paralel

Reply:
Yeah we could do that though now we pay the cost of a double array traversal (including within each file). I think we need to put this feature behind a flag, because I think in the default case this could cause issues, so only enable it if the user has asked for it because they have domains to check in bulk too.